File: WorldEditCUI/src/main/scala/jk_5/worldeditcui/network/packet/Packet.java
package jk_5.worldeditcui.network.packet;

/**
 * No description given
 *
 * @author jk-5
 */
public abstract class Packet {

    public Packet(){} //Keep the default constructor for reflection stuff

    private int minArguments;
    private int maxArguments;
    private String[] args;

    protected final void setNumberOfArguments(int min, int max){
        this.minArguments = min;
        this.maxArguments = max;
    }

    public final void processPacket(String[] args){
        this.args = args;
        if((this.maxArguments == this.minArguments && this.args.length != maxArguments) || (this.maxArguments != this.minArguments && (this.args.length > this.maxArguments || this.args.length < this.minArguments))){
            throw new PacketException("Received invalid number of arguments");
        }else{
            this.process();
        }
        this.process();
    }

    protected final int getIntArgument(int index){
        return Integer.parseInt(this.args[index]);
    }

    protected final double getDoubleArgument(int index){
        return Double.parseDouble(this.args[index]);
    }

    protected final String getStringArgument(int index){
        return this.args[index];
    }

    protected abstract void setupArguments();
    protected abstract void process() throws PacketException;
}

File: WorldEditCUI/src/main/scala/jk_5/worldeditcui/network/packet/Packet.scala
package jk_5.worldeditcui.network.packet

/**
 * No description given
 *
 * @author jk-5
 */
abstract class Packet(val minArgs: Int, val maxArgs: Int) {

  private var args: Array[String] = _

  final def processPacket(args: Array[String]){
    this.args = args
    if((this.maxArgs == this.minArgs && this.args.length != this.maxArgs) || (this.maxArgs != this.minArgs && (this.args.length > this.maxArgs || this.args.length < this.minArgs))){
      throw new PacketException("Received invalid number of arguments")
    }else{
      this.process()
    }
  }

  def process()

  def getInt(index: Int) = Integer.parseInt(this.args(index))
  def getDouble(index: Int) = java.lang.Double.parseDouble(this.args(index))
  def getString(index: Int) = this.args(index)
}

class PacketException(msg: String, cause: Throwable = null) extends RuntimeException(msg, cause)
